src/sugar.jl

```diff
@@ -1159,3 +1159,75 @@ grad
     )
     return nothing
 end
+
+"""
+    autodiff(
+        rmode::Union{ReverseMode,ReverseModeSplit},
+        f::Annotation,
+        dresult::Seed,
+        annotated_args...
+    )
+    
+Call [`autodiff_thunk`](@ref) in split mode, execute the forward pass, increment output adjoint with `dresult`, then execute the reverse pass.
+
+Useful for computing pullbacks / VJPs for functions whose output is not a scalar, or when the scalar seed is not 1.
+"""
+function autodiff(
+        rmode::Union{ReverseMode{ReturnPrimal}, ReverseModeSplit{ReturnPrimal}},
+        f::FA,
+        dresult::Seed{RT},
+        args::Vararg{Annotation, N},
+    ) where {ReturnPrimal, FA <: Annotation, RT, N}
+    rmode_split = Split(rmode)
+    RA = guess_activity(RT, rmode_split)
+    forward, reverse = autodiff_thunk(rmode_split, FA, RA, typeof.(args)...)
+    tape, result, shadow_result = forward(f, args...)
+    if RA <: Active
+        dinputs = only(reverse(f, args..., dresult.dval, tape))
+    else
+        Compiler.recursive_accumulate(shadow_result, dresult.dval)
+        dinputs = only(reverse(f, args..., tape))
+    end
+    if ReturnPrimal
+        return (dinputs, result)
+    else
+        return (dinputs,)
+    end
+end
+
+"""
+    autodiff(
+        rmode::Union{ReverseMode,ReverseModeSplit},
+        f::Annotation,
+        dresults::BatchSeed,
+        annotated_args...
+    )
+    
+Call [`autodiff_thunk`](@ref) in split mode, execute the forward pass, increment each output adjoint with the corresponding element from `dresults`, then execute the reverse pass.
+
+Useful for computing pullbacks / VJPs for functions whose output is not a scalar, or when the scalar seed is not 1.
+"""
+function autodiff(
+        rmode::Union{ReverseMode{ReturnPrimal}, ReverseModeSplit{ReturnPrimal}},
+        f::FA,
+        dresults::BatchSeed{B,RT},
+        args::Vararg{Annotation, N},
+    ) where {ReturnPrimal, B, FA <: Annotation, RT, N}
+    rmode_split_rightwidth = ReverseSplitWidth(Split(rmode), Val(B))
+    RA = batchify_activity(guess_activity(RT, rmode_split_rightwidth), Val(B))
+    forward, reverse = autodiff_thunk(rmode_split_rightwidth, FA, RA, typeof.(args)...)
+    tape, result, shadow_results = forward(f, args...)
+    if RA <: Active
+        dinputs = only(reverse(f, args..., dresults.dvals, tape))
+    else
+        foreach(shadow_results, dresults.dvals) do d0, d
```

Comment:
@vchuravy knows better than me here, but I don't think we should use foreach to ensure type stability, instead explicitly using either a generated function, or ntuple to go through?

Reply:
Do you have an example in mind where `foreach` would be unstable?